irc/handlers.go

```diff
@@ -889,6 +890,81 @@ func dlineHandler(server *Server, client *Client, msg ircmsg.IrcMessage, rb *Res
 	return killClient
 }
 
+// EXTJWT <target> [service_name]
+func extjwtHandler(server *Server, client *Client, msg ircmsg.IrcMessage, rb *ResponseBuffer) bool {
+	expireInSeconds := int64(30)
+
+	accountName := client.AccountName()
+	if accountName == "*" {
+		accountName = ""
+	}
+
+	claims := jwt.MapClaims{
+		"iss":     server.name,
+		"sub":     client.Nick(),
+		"account": accountName,
+		"umodes":  []string{},
+	}
+
+	if msg.Params[0] != "*" {
+		channel := server.channels.Get(msg.Params[0])
+		if channel == nil {
+			rb.Add(nil, server.name, "FAIL", "EXTJWT", "NO_SUCH_CHANNEL", client.t("No such channel"))
+			return false
+		}
+
+		claims["channel"] = channel.Name()
+		claims["joined"] = 0
+		claims["cmodes"] = []string{}
+		if channel.hasClient(client) {
+			joinTime := channel.ClientJoinTime(client)
+			if joinTime == nil {
+				// shouldn't happen, only in races
+				rb.Add(nil, server.name, "FAIL", "EXTJWT", "UNKNOWN_ERROR", client.t("Channel join time is inconsistent, JWT not generated"))
+				return false
+			}
+			claims["joined"] = joinTime.Unix()
+			claims["cmodes"] = channel.ClientModeStrings(client)
+		}
+	}
+
+	// we default to a secret of `*`. if you want a real secret setup a service in the config~
```

Comment:
What's going on here exactly?

Reply:
If you provide a service name (like `jitsi-call` or `acme-image-hosting`) then it'll use the secret defined in that service's config block. By default it just uses a secret of `*` because there's really no reason to have a default secret at all.

Reply:
Are we sure signing with a secret of `*` is correct? I found people saying that if there is no secret one should simply omit the signature: https://riptutorial.com/jwt/example/18558/unsigned-jwt

Reply:
Yes, we should not be mixing and matching with alg: none because of the very frequent issues that alg:none has caused in the past. I'm not sure whether using `*` in particular is correct, but if the spec recommends using none rather than hmac then it is a horrendous decision honestly.

Reply:
I'm not 100% clear on what's going on here but I have a really strong intuition that using a secret key of `*` is incorrect.

If I'm reading the spec correctly, we should be requiring the user to specify a default key of some kind (which should be a genuine secret and not `*`), and if they don't we should disable JWT (the command and the ISUPPORT token).

Should we ask for clarification on the spec PR?

Reply:
The spec should clarify this, yeah. There should not be any way to have a custom default secret, any custom secret should be provided with a service name. So we should probably only advertise JWT if there is at least one service configured.

Reply:
We got the [clarification](https://github.com/ircv3/ircv3-specifications/pull/341#issuecomment-616132786) we wanted:

1. The JWT header contains the signature algorithm, and the EXTJWT spec allows the use of any signature algorithm that can be verified. So I think what we have is fine. In the future, we could add, e.g., `RSAPrivateKey` to the `JWTServiceConfig`, and then the EXTJWT handler could sign with RSA instead if that field is populated.
2. We should in fact have a default secret (and EXTJWT should be disabled if one is not specified).

Reply:
1. So long as we never sign anything using the None algorithm, cool. Yeah could definitely add that key entry to the config in future no worries.
2. But that... doesn't answer our question. Should the default be _static_ (same on all nets, same everywhere, etc) or should the default be decided by the network and it's totally OK if clients can't actually verify it? Prawn's response indicates that this is intended for use by clients that don't actually know anything about the ircd setup as well so would having a configurable default actually make sense there or would it e.g. screw up clients trying to just decode data (for example do any client libraries out there intentionally not decode stuff if the user doesn't have the correct verification key?)

Reply:
>should the default be decided by the network and it's totally OK if clients can't actually verify it

I'm very confident that this is the correct answer. Clients do not parse/validate the JWT token themselves, they hand it off to a third party. That third party is assumed to hold the correct verification key. (The original use case for this is a closed ecosystem of an ircd and a Jitsi, controlled by the same operator.)

A "secret key" that is a constant value written into a specification just doesn't make sense.